src/operator/tensor/control_flow_op.h

```diff
@@ -0,0 +1,596 @@
+/*!
+ * Copyright (c) 2017 by Contributors
+ * \file control_flow.h
+ * \brief Function definitions of operators for controlling flow
+ */
+#ifndef MXNET_OPERATOR_TENSOR_CONTROL_FLOW_OP_H_
+#define MXNET_OPERATOR_TENSOR_CONTROL_FLOW_OP_H_
+
+#include <mxnet/operator_util.h>
+#include <vector>
+#include "../mshadow_op.h"
+#include "../mxnet_op.h"
+#include "../operator_common.h"
+#include "../elemwise_op_common.h"
+
+namespace mxnet {
+namespace op {
+
+/*! \brief return elements from x or y depending on condition
+ * The condition, x, and y have the same shape.
+ * The returned array is formed by elements from x or y
+ * depending on the elements of condition.
+ */
+template<int req>
+struct where;
+
+/*! \brief Choose elements from x or y depending on condition
+ * The condition, x, and y have the same shape.
+ * The returned array is formed by elements from x or y
+ * depending on the elements of condition.
+ * This should only be invoked by req_type=kNullOp.
+ */
+template<>
+struct where<kNullOp> {
```

Comment:
I think you can reuse ASSIGN_DISPATCH here (or write a similar macro), 
```
template<int req>
struct where{
MAP(...) {
    ASSIGN_DISPATCH(out[i], req, (0 != cond[i]? x[i] : y[i]));
}
}
```
It's much shorter